nixos/modules/services/networking/openconnect.nix

```
@@ -0,0 +1,154 @@
+{ config, lib, options, pkgs, ... }: with lib;
+let
+  cfg = config.networking.openconnect;
+  openconnect = cfg.package;
+  interfaceOptions = {
+    options = {
+      gateway = mkOption {
+        description = "Gateway server to connect to.";
+        example = "gateway.example.com";
+        type = types.str;
+      };
+
+      protocol = mkOption {
+        description = ''
+          Protocol to use. One of <literal>anyconnect</literal>,
+          <literal>array</literal>, <literal>nc</literal>,
+          <literal>pulse</literal>, <literal>gp</literal>,
+          <literal>f5</literal> and <literal>fortinet</literal>
+        '';
+        example = "anyconnect";
+        type = types.str;
+      };
+
+      user = mkOption {
+        description = "Username to authenticate with.";
+        example = "example-user";
+        type = types.nullOr types.str;
+      };
+
+      # Note: It does not make sense to provide a way to declaratively
+      # set an authentication cookie, because they have to be requested
+      # for every new connection and would only work once.
+      password = mkOption {
+        description = ''
+          Password to authenticate with. WARNING: To prevent this from being
+          stored world-readable in the Nix store, use
+          <code>passwordFile</code> instead. This is passed to
+          <code>openconnect</code> via the <code>--passwd-on-stdin</code> option.
+        '';
+        default = null;
+        type = types.nullOr types.str;
+      };
+
+      # Note
+      passwordFile = mkOption {
+        description = ''
+          File containing the password to authenticate with. This
+          is passed to <code>openconnect</code> via the
+          <code>--passwd-on-stdin</code> option.
+        '';
+        default = null;
+        example = "/var/lib/secrets/openconnect-passwd";
+        type = types.path;
+      };
+
+      certificate = mkOption {
+        description = ''
+          Certificate to authenticate with. Either a file name or
+          a PKCS#11 URL.
+        '';
+        default = null;
+        example = "/var/lib/secrets/openconnect_certificate.pem";
+        type = types.nullOr types.str;
+      };
+
+      privateKey = mkOption {
+        description = ''
+          Private key to authenticate with. Either a file name or
+          a PKCS#11 URL.
+        '';
+        example = "/var/lib/secrets/openconnect_private_key.pem";
+        default = null;
+        type = types.nullOr types.str;
```

```suggestion
        type = types.nullOr types.path;
```
the path type also allows strings as long as they begin with a `/`. if the module never interpolates the path without running it through `toString` first nothing will be copied into the store either (though the common pattern in nixos modules currently is to copy paths when given as path literals, not copying when they're strings)

---

The problem here is that this field can also be a PKCS#11 url (see https://www.infradead.org/openconnect/manual.html), so `types.path` would be too restrictive.

---

hmm. then how about `with types; nullOr (either path str)`?

---

Would that make any practical difference though? If I understand it correctly, that would essentially mean "either a string that starts with `/`, or a string"

---

it would insofar that the option docs state "path or string" as the type of the option, and it'd allow derivations without a bunch of wrapping (which could be beneficial for testing). ideally we'd want to have a `either path pkcs11` of sorts, but we don't know enough about pkcs11 to make a proper suggestion :/

(maybe with `pkcs11 = strMatching "pkcs11:.+" // { name = "PKCS#11 URI"; }`? it needn't do a complete validity check, just enough to catch most mistakes)

---

Cool! I didn't know that the type system was so flexible. I've pushed a new commit that uses the rudimentary pattern you suggested. (According to https://datatracker.ietf.org/doc/html/rfc7512#page-4 this should match against all valid PKCS#11 URIs.)